File: 0056-merge-intervals/0056-merge-intervals.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
const overlap = (x, y) => !(x[1] < y[0] || y[1] < x[0])

const merge = intervals => {
    intervals.sort((a, b) => a[0] - b[0]);
    let res = [intervals[0]];
    for (let i = 1; i < intervals.length; i++) {
        const curr = intervals[i];
        if (i === 0 || !overlap(curr, res[res.length - 1])) {
            res.push(curr);
        } else {
            res[res.length - 1][1] = Math.max(curr[1], res[res.length - 1][1]);
        }
    }
    return res;
};

File: 0056-merge-intervals/0056-merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        def sortBy(a):
            return a[0]
        intervals.sort(key=sortBy)
        merged = [intervals[0]]
        def overlap(a, b):
            return not (a[1] < b[0] or b[1] < a[0])
        left, right = 0, 1
        while right < len(intervals):
            a = merged[left]
            b = intervals[right]
            if overlap(a, b):
                merged[left] = [min(a[0], b[0]), max(a[1], b[1])]
                right += 1
            else:
                merged.append(b)
                right += 1
                left += 1
        return merged
    
    
    
    # [0, 2] [1, 4] [3, 5]
    # [0, 4] [3, 5]
    
    # [0, 2]